enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,543 @@
+- Introducing an OLM owned CRD forces operators to integrate with OLM at a much deeper level than writing to a resource that they already manage. Operators would be required to detect if it is being managed by OLM and begin writing to the API. By writing to an owned CR, the Operator does not require OLM specific code and can be run independently of OLM.
+
+Allowing operators to write to the status of any CR creates a potential scaling issue if enough CRs are on cluster. OLM will avoid these performance issues by introducing the Status Probe Operator (SPO) which will reconcile the `Probe` resource. If an operator chooses to write to their own CRs, SPO will create a pod that will keep the status of the `Probe` resource synced with the conditions in the CRs.
+
+>Note: Unfortunately, OLM will not support an operator communicating its status via cluster-scoped CRs. If an operator is scoped to a namespace but writes conditions to a cluster-scoped CR, OLM would be unable to identify which operator is communicating its status if multiple installs of the operator were present on cluster.
+
+### Opting into OLM Supported Conditions
+
+Operators may opt-in to "OLM Supported Conditions" by adding annotations to their owned CRD(s), where the key is the defined by OLM and the value is a boolean expression over condition types applied to the CR at `.status.conditions`. The "OLM Supported Condition" annotation key will adhere to the following format: `olm.operatorframework.io/condition.<ConditionName>`.
+
+Let's consider the case where an operator is upgradeable when the migrating condition on a CR is false, the "OLM Supported Condition" annotation should be `olm.operatorframework.io/condition.Upgradeable: "!Migrating"`. Alternatively, the operator could  indicate that it is upgradeable when both the `ReadyToGo` and `FinishedMigrating` conditions are true by adding the `olm.operatorframework.io/condition.Upgradeable: "FinishedMigrating && ReadyToGo"` annotation.
+
+### Example Workflow
+
+Let's examine the complete workflow for an operator that owns the `Foo` crd where:
+
+- The operator should not be upgraded when a `Foo` CR has the `Migrating` condition set to true.
+- OLM should surface any CR condition whose `Type` matches `BadConnectivity` or `UnhealthyDatabase` in the `Probe` resource.
+
+In this case, the operator's CRD should include the following annotations:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata: {
+  name: foo
+  annotations: {
+    "olm.operatorframework.io/condition.Upgradeable" : "!Migrating" # The operator is upgradeable when the CR is not being migrated.
+    "olm.operatorframework.io/condition.Important" : "BadConnectivity || UnhealthyDatabase" # The Important condition should only include the `OR` boolean operator.
+  }
+}
+```
+
+When this annotation is present on a CRD, OLM will append the following `Probe` CR to the `installPlan`:
```

Comment:
If the annotations define the probe when it's not provided in a bundle, can't we just generate probes dynamically at any time (not just in the InstallPlan)?

Reply:
Yes - but if an installPlan is a ledger of resources created as part of an installation, it seems like we should include the probe resource created for the operator within.

Reply:
That's a decent point. 

Devil's advocate: Do we add pods and replicasets for deployments we list in the installplan? We don't do this for `Operator` resources (which are created on-the-fly as well).

Real reasoning: InstallPlans are one-off installations. As a user, I would expect probe annotation changes to affect probes on-cluster. We've already seen pushback with CSV's and RBAC, where users have defined everything they need to in the CSV, but still need an InstallPlan to generate the RBAC for them; I see this problem as similar.